keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunner.kt

```diff
@@ -0,0 +1,250 @@
+package com.netflix.spinnaker.keel.actuation
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.anyStateful
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.constraints.ConstraintEvaluator
+import com.netflix.spinnaker.keel.api.constraints.ConstraintStatus.PENDING
+import com.netflix.spinnaker.keel.api.constraints.StatefulConstraintEvaluator
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactVetoes
+import com.netflix.spinnaker.keel.core.comparator
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+/**
+ * This class is responsible for running constraints and queueing them for approval.
+ * Approval into an environment is the responsibility of the [EnvironmentPromotionChecker].
+ */
+@Component
+class EnvironmentConstraintRunner(
+  private val repository: KeelRepository,
+  private val constraints: List<ConstraintEvaluator<*>>,
+  private val publisher: ApplicationEventPublisher
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val implicitConstraints: List<ConstraintEvaluator<*>> = constraints.filter { it.isImplicit() }
```

Comment:
Not for this PR, but I feel like we can extract these `constraints` and `evaluators` to a method / doing it in a different way..

Reply:
What do you mean/why would you want this to change? each of these statements is pretty short.

Reply:
I mean.. yeah it's pretty short but all of this `-` statements may be a bit confusing :) maybe it's just me :)

Reply:
list subtraction! so fun! :)